libraries/AP_BattMonitor/AP_BattMonitor_FuelLevel_Analog.cpp

```diff
@@ -0,0 +1,74 @@
+/*
+ * This file is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the
+ * Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This file is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Code by Charlie Johnson
+ */
+
+#include <AP_HAL/AP_HAL.h>
+#include "AP_BattMonitor_FuelLevel_Analog.h"
+
+extern const AP_HAL::HAL& hal;
+
+// Constructor
+AP_BattMonitor_FuelLevel_Analog::AP_BattMonitor_FuelLevel_Analog(AP_BattMonitor &mon, AP_BattMonitor::BattMonitor_State &mon_state, AP_BattMonitor_Params &params) :
+    AP_BattMonitor_Analog(mon, mon_state, params)
+{
+    _analog_source = hal.analogin->channel(_curr_pin);
+
+    // need to add check
+    _state.healthy = false;
+
+    // create a low pass filter
+    _voltage_filter.set_cutoff_frequency(0.3f);
+}
+
+/*
+  read - read the "voltage" and "current"
+*/
+void AP_BattMonitor_FuelLevel_Analog::read()
+{
+    if (!_analog_source->set_pin(_curr_pin)) {
+        _state.healthy = false;
+        return;
+    }
+    _state.healthy = true;
+
+    // get a dt and a timestamp
+    const uint32_t tnow = AP_HAL::micros();
+    const uint32_t dt_us = tnow - _state.last_time_micros;
+
+    // get votlage from an ADC pin and filter it
+    const float voltage = _analog_source->voltage_average();
+    const float filtered_voltage = _voltage_filter.apply(voltage, float(dt_us) / 1000.0);
+
+    // output the ADC voltage to the voltage field for easier calibration of sensors
+    _state.voltage = voltage;
+    _state.current_amps = filtered_voltage;
+
+    // this driver assumes that CAPACITY is set to tank volume in millilitres.
+    // the empty voltage is _volt_offset
+    // the full voltage is _volt_multiplier
+    const float &voltage_used = (uint32_t(_params._options.get()) & uint32_t(AP_BattMonitor_Params::Options::FuelLevel_Analog_LPF))?filtered_voltage:voltage;
+    const float fuel_level_ratio = (voltage_used - _volt_offset) / (_volt_multiplier - _volt_offset);
```

Comment:
Why are we subtracting the volt offset from the multiplier? That seems counter intutive to me, I was expecting this to be: `(voltage_used - _volt_offset) * _volt_multiplier`

Reply:
I chose to go with the approach of setting both points and figuring out the slope rather than trying to fuss with setting a slope as a user. Both parameters can be set by reading the values from the GCS with empty and full tanks.

Reply:
It just seems odd to me that fine tuning your empty point will actually throw off any scaling correction you've done/know. It also just stands out as being done very different then the analog battery current monitor which does it as I outlined above, and as a user familiar with the analog battery I would have expected this to look more like that.

Reply:
Particularly to do it as you've implemented I think needs a lot more in terms of additions to the parameter documentation.